Raven.Database/Raft/ClusterManager.cs

```diff
@@ -53,71 +111,73 @@ private void OnRaftEngineStateChanged(RaftEngineState state)
-                timer?.Dispose();
+                SaflyDisposeOfTimer();
                 timer = new Timer(NewLeaderTimerCallback,null, period, period); 
             }
             else
             {
                 //If this node was the leader before, dispose of the timer.
-                timer?.Dispose();
+                SaflyDisposeOfTimer();
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SaflyDisposeOfTimer()
+        {
+            var timerSnapshot = timer;
+            timerSnapshot?.Dispose();
+            timer = null;
+        }
+
+        private static int newLeaderTimerCallbackLocker;
         private void NewLeaderTimerCallback(object state)
         {
-            //This method should never be invoked from non-leader
-            //Might happen if the timer and the state change event happen at the same time.
-            if (Engine.State != RaftEngineState.Leader)
-            {
-                if(Log.IsDebugEnabled)
-                    Log.Debug($"NewLeaderTimerCallback invoked from non-leader node, actual state:{Engine.State}");
+            //preventing multiple calls to the callback
+            if (Interlocked.CompareExchange(ref newLeaderTimerCallbackLocker, 1, 0) == 1)
                 return;
-            }
-            var databasesNames = GetDatabasesNames();
-            var databaseToLastModified = new Dictionary<string,Tuple<DateTime,string>>();
-            foreach (var databaseName in databasesNames)
+            try
             {
-                try
+                //This method should never be invoked from non-leader
+                //Might happen if the timer and the state change event happen at the same time.
+                if (Engine.State != RaftEngineState.Leader)
                 {
-                    Task<DocumentDatabase> databaseTask;
-                    if (DatabasesLandlord.TryGetOrCreateResourceStore(databaseName, out databaseTask) == false)
-                        continue;
-                    var database = databaseTask.Result;
-                    JsonDocument lastDoc = null;
-                    database.TransactionalStorage.Batch(action =>
+                    if(Log.IsDebugEnabled)
+                        Log.Debug($"NewLeaderTimerCallback invoked from non-leader node, actual state:{Engine.State}");
+                    return;
+                }
+                var databaseToLastModified = new Dictionary<string, LastModificationTimeAndTransactionalId>();
+                DatabasesLandlord.ForAllDatabases(db => 
+                {
+                    try
                     {
-                        lastDoc = action.Documents.GetDocumentsByReverseUpdateOrder(0, 1).FirstOrDefault();                        
-
-                    });
-                    if (lastDoc == null)
+                        var databaseName = db.Name;
+                        JsonDocument lastDoc = null;
```

Comment:
You are assuming here that I'm replicating to the same db name.
Need to verify that this is indeed always the case in clusters (pretty sure it is, mind)